test/tests/pbs_qstat.py

```diff
@@ -0,0 +1,72 @@
+# coding: utf-8
+
+# Copyright (C) 1994-2018 Altair Engineering, Inc.
+# For more information, contact Altair at www.altair.com.
+#
+# This file is part of the PBS Professional ("PBS Pro") software.
+#
+# Open Source License Information:
+#
+# PBS Pro is free software. You can redistribute it and/or modify it under the
+# terms of the GNU Affero General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# For a copy of the commercial license terms and conditions,
+# go to: (http://www.pbspro.com/UserArea/agreement.html)
+# or contact the Altair Legal Department.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from ptl.utils.pbs_testsuite import *
+
+
+class QstatTest(PBSTestSuite):
+
+    """
+    This test suite checks whether qstat sends the
+    status request to the correct server or not for
+    requests like job_id.serverA@serverB
+    """
+
+    def setUp(self):
+        if len(self.servers) != 2:
+            self.skipTest("This test needs two servers")
+        self.server1 = self.servers.values()[0]
+        self.server2 = self.servers.values()[1]
+        a = {'scheduling': 'false', 'flatuid': 'true'}
+        self.server1.manager(MGR_CMD_SET, SERVER, a)
+        self.server2.manager(MGR_CMD_SET, SERVER, a)
+        self.du.set_pbs_config(self.server.client, confs={'PBS_SERVER': ''})
+
+    def test_qstat_req_server(self):
+        j = Job(TEST_USER)
+        actualclient = self.server.client
```

Comment:
I don't think you need this client toggling. PBSTestSuite.setup() will set proper client to whatever given on pbs_bechpress -p client

Reply:
I think this will be required because in the setup I am making the PBS_SERVER empty in pbs_conf of client. So I have to toggle the client so that the job can be submitted and moved.

Reply:
I don't think you need to unset PBS_SERVER entry in pbs.conf on client, because if job id has server name in it then qstat (and as far as I know almost all PBS client command) will use that as server to connect rather than PBS_SERVER in pbs.conf.
So even it you keep pbs.conf as it is on client (or set it to server1 so all default request goes to server1) and remove this toggling then also your test should run fine.
(See https://github.com/PBSPro/pbspro/blob/master/src/lib/Libcmds/get_server.c#L142 and https://github.com/PBSPro/pbspro/blob/master/src/cmds/qstat.c#L2628)

Reply:
@hirenvadalia - The issue that is being fixed is that qstat is not sending the status request to the proper server. I had explained @Jitendra-Rajak in an internal discussion as to how we can reproduce this particular issue by not having PBS_SERVER=server1 in client's pbs.conf.
The test case is passing with the latest change as PBS_SERVER is set to server1 in pbs.conf on the client machine and the actual issue itself is not getting reproduced. With the last change, I think the test case will pass on builds without fix as well, @Jitendra-Rajak - can confirm.
This is because the comparison on line 153 (https://github.com/PBSPro/pbspro/blob/b8eb6ffb5eb68009aee1c87837f5325022326574/src/lib/Libcmds/get_server.c#L153) in get_server.c will fail and get_server() will not return on line 157.
It is my opinion that we should reverse the last change.

Reply:
Yes @hirenvadalia, I was not able to reproduce the issue with having PBS_SERVER in client's pbs.conf. So after that @prakashcv13 and I were able to reproduce the issue by making PBS_SERVER empty in client's pbs.conf. 
That is why in the setup I was making PBS_SERVER empty so that the test follows the actual scenario in which the bug was logged. I think reversing the change should be the perfect solution.